lib/src/st2110/pipeline/st40_pipeline_tx.c

```diff
@@ -0,0 +1,563 @@
+  return 0;
+}
+
+static int tx_st40p_frame_done(void* priv, uint16_t frame_idx,
+                               struct st40_tx_frame_meta* meta) {
+  struct st40p_tx_ctx* ctx = priv;
+  int ret;
+  struct st40p_tx_frame* framebuff = &ctx->framebuffs[frame_idx];
+
+  mt_pthread_mutex_lock(&ctx->lock);
+  if (ST40P_TX_FRAME_IN_TRANSMITTING == framebuff->stat) {
+    ret = 0;
+    framebuff->stat = ST40P_TX_FRAME_FREE;
+    dbg("%s(%d), done_idx %u\n", __func__, ctx->idx, frame_idx);
+  } else {
+    ret = -EIO;
+    err("%s(%d), err status %d for frame %u\n", __func__, ctx->idx, framebuff->stat,
+        frame_idx);
+  }
+  mt_pthread_mutex_unlock(&ctx->lock);
+
+  struct st40_frame_info* frame = &framebuff->frame;
+  frame->tfmt = meta->tfmt;
+  frame->timestamp = meta->timestamp;
+  frame->epoch = meta->epoch;
+  frame->rtp_timestamp = meta->rtp_timestamp;
+
+  if (ctx->ops.notify_frame_done) { /* notify app which frame done */
+    ctx->ops.notify_frame_done(ctx->ops.priv, frame);
+  }
+
+  /* notify app can get frame */
+  tx_st40p_notify_frame_available(ctx);
+
+  return ret;
+}
+
+static int tx_st40p_create_transport(struct mtl_main_impl* impl, struct st40p_tx_ctx* ctx,
+                                     struct st40p_tx_ops* ops) {
+  int idx = ctx->idx;
+  struct st40_tx_ops ops_tx;
+
+  memset(&ops_tx, 0, sizeof(ops_tx));
+  ops_tx.name = ops->name;
+  ops_tx.priv = ctx;
+  ops_tx.num_port = RTE_MIN(ops->port.num_port, MTL_SESSION_PORT_MAX);
+  ops_tx.payload_type = ops->port.payload_type;
+  ops_tx.ssrc = ops->port.ssrc;
+  for (int i = 0; i < ops_tx.num_port; i++) {
+    memcpy(ops_tx.dip_addr[i], ops->port.dip_addr[i], MTL_IP_ADDR_LEN);
+    snprintf(ops_tx.port[i], MTL_PORT_MAX_LEN, "%s", ops->port.port[i]);
+    ops_tx.udp_src_port[i] = ops->port.udp_src_port[i];
+    ops_tx.udp_port[i] = ops->port.udp_port[i];
+  }
+  if (ops->flags & ST40P_TX_FLAG_USER_P_MAC) {
+    memcpy(&ops_tx.tx_dst_mac[MTL_SESSION_PORT_P][0],
+           &ops->tx_dst_mac[MTL_SESSION_PORT_P][0], MTL_MAC_ADDR_LEN);
+    ops_tx.flags |= ST40_TX_FLAG_USER_P_MAC;
+  }
+  if (ops->flags & ST40P_TX_FLAG_USER_R_MAC) {
+    memcpy(&ops_tx.tx_dst_mac[MTL_SESSION_PORT_R][0],
+           &ops->tx_dst_mac[MTL_SESSION_PORT_R][0], MTL_MAC_ADDR_LEN);
+    ops_tx.flags |= ST40_TX_FLAG_USER_R_MAC;
+  }
+  if (ops->flags & ST40P_TX_FLAG_DEDICATE_QUEUE)
+    ops_tx.flags |= ST40_TX_FLAG_DEDICATE_QUEUE;
+  if (ops->interlaced) {
+    err("%s(%d), interlaced mode not supported\n", __func__, ctx->idx);
+    return -EINVAL;
+  }
+  ops_tx.interlaced = false;
+  ops_tx.fps = ops->fps;
+
+  ops_tx.framebuff_cnt = ops->framebuff_cnt;
+  ops_tx.type = ST40_TYPE_FRAME_LEVEL;
+  ops_tx.get_next_frame = tx_st40p_next_frame;
+  ops_tx.notify_frame_done = tx_st40p_frame_done;
+
+  ctx->transport = st40_tx_create(impl, &ops_tx);
+  if (!ctx->transport) {
+    err("%s(%d), transport create fail\n", __func__, idx);
+    return -EIO;
+  }
+
+  struct st40p_tx_frame* frames = ctx->framebuffs;
```

Comment:
Please move the declaration to the top

Reply:
good good. It was copied without much thought from different part of MTL, and now when we review this code we can see how many "innovative stylistic choices" where made before

Reply:
moved all declaration to the top